src/jit/lsraxarch.cpp

```diff
@@ -1372,166 +1429,30 @@ void Lowering::TreeNodeInfoInitCall(GenTreeCall* call)
-                    remainingSlots--;
-                    regCount++;
-                }
-            }
+        if (argNode->OperGet() == GT_FIELD_LIST)
+        {
+            assert(varTypeIsStruct(argNode) || curArgTabEntry->isStruct);
 
-            short internalIntCount = 0;
-            if (remainingSlots > 0)
+            unsigned eightbyte = 0;
+            for (GenTreeFieldList* entry = argNode->AsFieldList(); entry != nullptr; entry = entry->Rest())
             {
-#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
-                // This TYP_STRUCT argument is also passed in the outgoing argument area
-                // We need a register to address the TYP_STRUCT
-                internalIntCount = 1;
-#else  // FEATURE_UNIX_AMD64_STRUCT_PASSING
-                // And we may need 2
-                internalIntCount           = 2;
-#endif // FEATURE_UNIX_AMD64_STRUCT_PASSING
-            }
-            argNode->gtLsraInfo.internalIntCount = internalIntCount;
+                const regNumber argReg = eightbyte == 0 ? curArgTabEntry->regNum : curArgTabEntry->otherRegNum;
+                TreeNodeInfoInitPutArgReg(entry->Current()->AsUnOp(), argReg, *info, isVarArgs, &callHasFloatRegArgs);
 
-#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
-            if (argNode->gtOper == GT_PUTARG_REG)
-            {
-                argNode->gtLsraInfo.setDstCandidates(l, argMask);
-                argNode->gtLsraInfo.setSrcCandidates(l, argMask);
+                eightbyte++;
             }
-#endif // FEATURE_UNIX_AMD64_STRUCT_PASSING
         }
         else
+#endif // FEATURE_UNIX_AMD64_STRUCT_PASSING
         {
-            argNode->gtLsraInfo.setDstCandidates(l, argMask);
-            argNode->gtLsraInfo.setSrcCandidates(l, argMask);
-        }
-
-        // To avoid redundant moves, have the argument child tree computed in the
-        // register in which the argument is passed to the call.
-        if (argNode->gtOper == GT_PUTARG_REG)
-        {
-            argNode->gtOp.gtOp1->gtLsraInfo.setSrcCandidates(l, l->getUseCandidates(argNode));
-        }
-
-#if FEATURE_VARARG
-        // In the case of a varargs call, the ABI dictates that if we have floating point args,
-        // we must pass the enregistered arguments in both the integer and floating point registers.
-        // Since the integer register is not associated with this arg node, we will reserve it as
-        // an internal register so that it is not used during the evaluation of the call node
-        // (e.g. for the target).
-        if (call->IsVarargs() && varTypeIsFloating(argNode))
-        {
-            regNumber targetReg = compiler->getCallArgIntRegister(argReg);
-            info->setInternalIntCount(info->internalIntCount + 1);
-            info->addInternalCandidates(l, genRegMask(targetReg));
+            TreeNodeInfoInitPutArgReg(argNode->AsUnOp(), curArgTabEntry->regNum, *info, isVarArgs,
+                                      &callHasFloatRegArgs);
         }
```

Comment:
It might be a good idea to generate asm diffs on unix amd64 to make sure there are zero asm diffs.